lib/davinci_dtr_test_kit/dtr_questionnaire_response_validation.rb

```diff
@@ -1,30 +1,77 @@
-require_relative 'fixtures'
+# frozen_string_literal: true
 
 module DaVinciDTRTestKit
   module DTRQuestionnaireResponseValidation
-    include Fixtures
-
     CQL_EXPRESSION_EXTENSIONS = [
       'http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-initialExpression',
       'http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-calculatedExpression',
       'http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-candidateExpression'
     ].freeze
 
-    def validate_questionnaire_pre_population(questionnaire_response, test_id)
-      # Requirements:
-      #  - Prior to exposing the draft QuestionnaireResponse to the user for completion and/or review, the DTR client
-      #    SHALL execute all CQL necessary to resolve the initialExpression, candidateExpression and
-      #    calculatedExpression extensions found in the Questionnaire for any enabled elements.
-      #  - All items that are pre-populated (whether by the payer in the initial QuestionnaireResponse provided in the
-      #    questionnaire package, or from data retrieved from the EHR) SHALL have their origin.source set to ‘auto’.
-      #
-      # Note that in the questionnaire fixture, all cql expression elements are enabled, so we don't filter
-      template_questionnaire_response = find_questionnaire_response_for_test_id(test_id)
-      raise "missing QuestionnaireResponse template for test #{test_id}" unless template_questionnaire_response.present?
-
-      questionnaire = find_questionnaire_instance_for_test_id(test_id)
-      raise "missing Questionnaire for test #{test_id}" unless questionnaire.present?
+    def check_is_questionnaire_response(questionnaire_response_json)
+      assert_valid_json(questionnaire_response_json)
+      questionnaire_response = begin
+        FHIR.from_contents(questionnaire_response_json)
+      rescue StandardError
+        nil
+      end
+
+      assert questionnaire_response.present?, 'The QuestionnaireResponse is not a recognized FHIR object'
+      assert_resource_type(:questionnaire_response, resource: questionnaire_response)
+    end
+
+    def verify_basic_conformance(questionnaire_response_json)
+      check_is_questionnaire_response(questionnaire_response_json)
+      assert_valid_resource(resource: FHIR.from_contents(questionnaire_response_json),
+                            profile_url: 'http://hl7.org/fhir/us/davinci-dtr/StructureDefinition/dtr-questionnaireresponse|2.0.1')
+    end
+
+    # This only checks answers in the questionnaire response, meaning it does not catch missing answers
+    def check_origin_sources(questionnaire_items, response_items, expected_overrides: [])
```

Comment:
This method has a lot of similarities with the check_item_prepopulation method, except that the specific values of the answers aren't checked (I think). Is there a reason to keep them separate, or could we have a single set of logic with a flag that turns specific value checking on or off?

Reply:
See PR description:

> This adds some duplication of logic and error messages between this and the pre-population validation. I tried for a while to refactor things to reduce this duplication, but I was struggling to reach a solution that was adequately clear and simple, so I ultimately decided the duplication was preferable.

I really didn't want to add flags to the existing prepopulation validation stuff (`validate_questionnaire_pre_population`) because it's clunky and messy enough as it is. But I'll take another look at just reusing the `check_item_prepopulation` method.